core/iwasm/common/wasm_shared_memory.c

```diff
@@ -486,8 +503,8 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
     wasm_runtime_free(wait_node);
 
     /* Release wait info if no wait nodes attached */
-    removed_from_map = map_remove_wait_info(wait_map, wait_info, address);
     os_mutex_unlock(&wait_info->wait_list_lock);
+    removed_from_map = map_remove_wait_info(wait_map, wait_info, address);
     if (removed_from_map)
         destroy_wait_info(wait_info);
```

Comment:
We can probably move this line inside `map_remove_wait_info` and rename `map_remove_wait_info` to `release_wait_info`. I made the reverse change in a previous PR, but with your modifications it makes more sense to restore it as it was before.

Reply:
Fixed